### PR TITLE
[CAMEL-14468] Add support for classification in camel-weka

### PR DESCRIPTION
**Pull request by tdiesler**

### CODE REVIEW

**Comment by tdiesler** [comment]
Closing, ascii doc says that params are applicable both for producer and consumer.

**Comment by tdiesler** [comment]
Its a docs issue, that can perhaps be dealt with later

File: components/camel-weka/src/main/java/org/apache/camel/component/weka/WekaEndpoint.java [hunk 1]
```diff
@@ -16,22 +16,46 @@
...
+import weka.core.converters.ArffLoader;
+import weka.core.converters.CSVLoader;
+import weka.core.converters.Loader;
 
 /**
  * The camel-weka component provides Data Mining functionality through Weka.
  */
-@UriEndpoint(firstVersion = "3.1.0", scheme = "weka", title = "Weka",
-        syntax = "weka:cmd?options", producerOnly = true, label = "data mining")
+@UriEndpoint(firstVersion = "3.1.0", scheme = "weka", title = "Weka", syntax = "weka:cmd?options", label = "data mining")
```
Comment (davsclaus) [inline components/camel-weka/src/main/java/org/apache/camel/component/weka/WekaEndpoint.java]: No ?option in syntax. And label should be a single word, datamining or something as a category label.

File: components/camel-weka/src/main/java/org/apache/camel/component/weka/WekaConsumer.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+    public WekaConfiguration getConfiguration() {
+        return getEndpoint().getConfiguration();
+    }
+    
+    @Override
+    protected void doStart() throws Exception {
+        super.doStart();
+        
+        Thread thread = new Thread(new Runnable() {
```
Comment (davsclaus) [inline components/camel-weka/src/main/java/org/apache/camel/component/weka/WekaConsumer.java]: Its preferred to create thread via Camel. See CamelContext ExecutorServiceStrategy which has APIs for creating threads. Also you should stop the thread in doStop.

File: components/camel-weka/src/main/java/org/apache/camel/component/weka/WekaConsumer.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+            
+            @Override
+            public void run() {
+                
+                WekaEndpoint endpoint = getEndpoint();
+                Exchange exchange = endpoint.createExchange();
+                
+                Command cmd = getConfiguration().getCommand();
+                if (Command.read != cmd) {
```
Comment (davsclaus) [inline components/camel-weka/src/main/java/org/apache/camel/component/weka/WekaConsumer.java]: Check for invalid command in the component or the createConsumer method. Doing this here is too late.

File: components/camel-weka/src/main/java/org/apache/camel/component/weka/WekaEndpoint.java [hunk 4]
```diff
@@ -44,14 +68,18 @@ public WekaEndpoint(String uri, WekaComponent component, WekaConfiguration confi
...
     public WekaComponent getComponent() {
-        return (WekaComponent)super.getComponent();
+        return (WekaComponent) super.getComponent();
     }
 
     @Override
-    public Consumer createConsumer(Processor processor) throws Exception {
-        throw new UnsupportedOperationException();
+    public Consumer createConsumer(Processor processor) {
+        return new WekaConsumer(this, processor);
```
Comment (davsclaus) [inline components/camel-weka/src/main/java/org/apache/camel/component/weka/WekaEndpoint.java]: You need to call configureConsumer too, see the other components

File: components/camel-weka/src/main/java/org/apache/camel/component/weka/WekaEndpoint.java [hunk 5]
```diff
@@ -64,11 +92,245 @@ public boolean isSingleton() {
...
+        // Load the Model
+        
+        if (loadFrom != null) {
+            
+            Classifier cl = dataset
+                    .loadClassifier(new ModelLoader(loadFrom))
+                    .getClassifier();
+            
+            AssertState.notNull(cl, "Cannot load the classifier from: " + loadFrom);
+            LOG.info("{}", cl);
```
Comment (davsclaus) [inline components/camel-weka/src/main/java/org/apache/camel/component/weka/WekaEndpoint.java]: LOG.debug

File: components/camel-weka/src/main/java/org/apache/camel/component/weka/WekaEndpoint.java [hunk 6]
```diff
@@ -64,11 +92,245 @@ public boolean isSingleton() {
...
+                    dataset.pop(dsname);
+                }
+                
+                // Train with current instances
+                dataset.evaluateModel();
+            }
+            
+            Classifier cl = dataset.getClassifier();
+            AssertState.notNull(cl, "Model command requires 'load' or 'apply'");
+            LOG.info("{}", cl);
```
Comment (davsclaus) [inline components/camel-weka/src/main/java/org/apache/camel/component/weka/WekaEndpoint.java]: debug logging

File: components/camel-weka/src/main/java/org/apache/camel/component/weka/WekaEndpoint.java [hunk 7]
```diff
@@ -64,11 +92,245 @@ public boolean isSingleton() {
...
+            
+            Evaluation ev = dataset.getEvaluation();
+            LOG.info("{}", ev.toSummaryString());
```
Comment (davsclaus) [inline components/camel-weka/src/main/java/org/apache/camel/component/weka/WekaEndpoint.java]: Is this needed to be logged at info level? And is this per exchange which then can get noisy

File: components/camel-weka/src/main/java/org/apache/camel/component/weka/WekaEndpoint.java [hunk 8]
```diff
@@ -64,11 +92,245 @@ public boolean isSingleton() {
...
+                
+                GenericFile<?> file = (GenericFile<?>) body;
+                AssertState.isFalse(file.isDirectory(), "Directory not supported: " + file);
+                String absolutePath = file.getAbsoluteFilePath();
+                dataset = Dataset.create(absolutePath);
+                
+            } else if (body instanceof URL) {
+                
+                URL url = (URL) body;
+                Instances instances = readInternal(url.openStream());
```
Comment (davsclaus) [inline components/camel-weka/src/main/java/org/apache/camel/component/weka/WekaEndpoint.java]: Can the stream not leak? eg need to close it after use

File: components/camel-weka/src/main/java/org/apache/camel/component/weka/WekaEndpoint.java [hunk 9]
```diff
@@ -64,11 +92,245 @@ public boolean isSingleton() {
...
+                dataset = Dataset.create(instances);
+            }
+        }
+        
+        AssertState.notNull(dataset, "Cannot obtain dataset from body: " + body);
+        return dataset;
+    }
+
+    // https://github.com/tdiesler/nessus-weka/issues/11
+    private static Instances readInternal(InputStream input) {
```
Comment (davsclaus) [inline components/camel-weka/src/main/java/org/apache/camel/component/weka/WekaEndpoint.java]: Can the stream not leak? eg need to close it after use

File: components/camel-weka/src/test/java/org/apache/camel/component/weka/DecisionTreeTest.java [hunk 10]
```diff
@@ -0,0 +1,124 @@
+package org.apache.camel.component.weka;
```
Comment (davsclaus) [inline components/camel-weka/src/test/java/org/apache/camel/component/weka/DecisionTreeTest.java]: Add license headers for missing files.

**Comment by davsclaus** [comment]
btw great work Thomas on new components, great to see you keep adding new ones about these new emerging techs

File: components/camel-weka/src/main/java/org/apache/camel/component/weka/WekaEndpoint.java (hunk 1 above)
Comment (tdiesler) [reply components/camel-weka/src/main/java/org/apache/camel/component/weka/WekaEndpoint.java]: Done

File: components/camel-weka/src/main/java/org/apache/camel/component/weka/WekaConsumer.java (hunk 3 above)
Comment (tdiesler) [reply components/camel-weka/src/main/java/org/apache/camel/component/weka/WekaConsumer.java]: Done

File: components/camel-weka/src/main/java/org/apache/camel/component/weka/WekaEndpoint.java (hunk 4 above)
Comment (tdiesler) [reply components/camel-weka/src/main/java/org/apache/camel/component/weka/WekaEndpoint.java]: Done

File: components/camel-weka/src/main/java/org/apache/camel/component/weka/WekaEndpoint.java (hunk 6 above)
Comment (tdiesler) [reply components/camel-weka/src/main/java/org/apache/camel/component/weka/WekaEndpoint.java]: Done

File: components/camel-weka/src/main/java/org/apache/camel/component/weka/WekaEndpoint.java (hunk 5 above)
Comment (tdiesler) [reply components/camel-weka/src/main/java/org/apache/camel/component/weka/WekaEndpoint.java]: Done

File: components/camel-weka/src/main/java/org/apache/camel/component/weka/WekaEndpoint.java (hunk 7 above)
Comment (tdiesler) [reply components/camel-weka/src/main/java/org/apache/camel/component/weka/WekaEndpoint.java]: Done

File: components/camel-weka/src/main/java/org/apache/camel/component/weka/WekaEndpoint.java (hunk 8 above)
Comment (tdiesler) [reply components/camel-weka/src/main/java/org/apache/camel/component/weka/WekaEndpoint.java]: Well spotted, thanks.

File: components/camel-weka/src/main/java/org/apache/camel/component/weka/WekaEndpoint.java (hunk 9 above)
Comment (tdiesler) [reply components/camel-weka/src/main/java/org/apache/camel/component/weka/WekaEndpoint.java]: Caller resposibility

**Comment by tdiesler** [comment]
The above should all be done

**Pull request merged**